staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -132,21 +137,89 @@ func BuildAndRegisterAggregator(downloader *Downloader, delegationTarget server.
 	}
 	err = s.openAPIVersionedService.RegisterOpenAPIVersionedService("/openapi/v2", pathHandler)
 	if err != nil {
-		return nil, err
+		klog.Fatalf("Failed to register versioned open api spec for root: %v", err)
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.OpenAPIV3) {
+		s.openAPIV3VersionedService, err = handler3.NewOpenAPIService(nil)
+		if err != nil {
+			return nil, err
+		}
+		err = s.openAPIV3VersionedService.RegisterOpenAPIV3VersionedService("/openapi/v3", pathHandler)
+		if err != nil {
+			return nil, err
+		}
+
+		i = 1
+		for delegate := delegationTarget; delegate != nil; delegate = delegate.NextDelegate() {
+			handler := delegate.UnprotectedHandler()
+			if handler == nil {
+				continue
+			}
+
+			apiServiceName := fmt.Sprintf(localDelegateChainNamePattern, i)
+			s.UpdateOpenAPIV3(*downloader, apiServiceName, handler)
+			i++
+		}
 	}
 
 	return s, nil
 }
 
+func (s *specAggregator) UpdateOpenAPIV3(downloader Downloader, apiServiceName string, handler http.Handler) error {
```

Comment:
iit's not apparent what an `Update*` func does. Is it adding another handler? Name it properly.

Reply:
Renamed to AddUpdateAPIServiceForOpenAPIV3 and added description